File: android/src/main/kotlin/com/pushly/pushsdk/delegates/UserProfileDelegate.kt
package com.pushly.pushsdk.delegates

import android.content.Context
import com.pushly.android.PushSDK
import com.pushly.pushsdk.FlutterResponder
import com.pushly.pushsdk.PNLogs
import com.pushly.pushsdk.mappers.toMap
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class UserProfileDelegate(
    private val context: Context,
    internal val channel: MethodChannel,
    private val binaryMessenger: BinaryMessenger,
) : FlutterResponder(), MethodChannel.MethodCallHandler {

    companion object {
        fun create(context: Context, binaryMessenger: BinaryMessenger): UserProfileDelegate {
            return UserProfileDelegate(context, MethodChannel(binaryMessenger, "PushSDK#UserProfile"), binaryMessenger)
        }
    }

    init {
        initialize(context, channel, binaryMessenger)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        PNLogs.debug("Android Flutter bridge received call: ${call.method} with args: ${call.arguments}")
        when (call.method) {
            "get" -> {
                val profile = PushSDK.UserProfile.toMap()
                PNLogs.debug("Sending UserProfile#get with value: $profile")
                success(result, profile)
            }
            "setExternalId" -> {
                PushSDK.UserProfile.externalId = call.argument("externalId")
                success(result)
            }
            "append" -> {
                PushSDK.UserProfile.append(key = call.argument<String>("key")!!, values = call.argument<List<Any>>("values")!!)
                success(result)
            }
            "set" -> {
                PushSDK.UserProfile.set(key = call.argument<String>("key")!!, value = call.argument<Any>("value")!!)
                success(result)
            }
            "setData" -> {
                PushSDK.UserProfile.set(call.argument<HashMap<String, Any>>("data")!!)
                success(result)
            }
            "remove" -> {
                PushSDK.UserProfile.remove(key = call.argument<String>("key")!!, values = call.argument<List<Any>>("values")!!)
                success(result)
            }
            "trackActivity" -> {
                PushSDK.UserProfile.trackActivity(
                    name = call.argument<String>("name")!!,
                    tags = call.argument<List<String>>("tags")!!)
                success(result)
            }
            "requestUserDeletion" -> {
                PushSDK.UserProfile.requestUserDeletion()
                success(result)
            }
            else -> {
                notImplemented(result)
                return
            }
        }

    }
}

File: ios/Classes/Delegates/UserProfileDelegate.swift
import Pushly
import Flutter
import UIKit

internal class UserProfileDelegate: NSObject, FlutterPlugin {
    
    private let channel: FlutterMethodChannel
    private let binaryMessenger: FlutterBinaryMessenger
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let instance = UserProfileDelegate(binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: instance.channel)
    }
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        self.binaryMessenger = binaryMessenger
        channel = FlutterMethodChannel(name: "PushSDK#UserProfile", binaryMessenger: binaryMessenger)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        PNLogs.debug("iOS Flutter bridge received method call: \(call.method) with args \(String(describing: call.arguments))")
        let args: [String: Any?] = (call.arguments as? Dictionary<String, Any?>) ?? [:]
        
        switch call.method {
        case "get":
            let profile = PushSDK.UserProfile.dict()
            PNLogs.debug("Sending UserProfile#get with value: \(profile)")
            result(profile)
        case "setExternalId":
            PushSDK.UserProfile.externalId = args["externalId"] as! String
            result(nil)
        case "append":
            PushSDK.UserProfile.append(args["values"] as! [Any], to: args["key"] as! String)
            result(nil)
        case "set":
            PushSDK.UserProfile.set(args["value"]!!, forKey: args["key"] as! String)
            result(nil)
        case "setData":
            PushSDK.UserProfile.set(args["data"] as! [String: Any])
            result(nil)
        case "remove":
            PushSDK.UserProfile.remove(args["values"] as! [Any], from: args["key"] as! String)
            result(nil)
        case "trackActivity":
            if let tags = args["tags"] as? [String] {
                PushSDK.UserProfile.trackActivity(name: args["name"] as! String, withTags: tags)
            } else {
                PushSDK.UserProfile.trackActivity(name: args["name"] as! String)
            }
            result(nil)
        case "requestUserDeletion":
            PushSDK.UserProfile.requestUserDeletion()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

